text/0000-unnamed-fields.md

```diff
@@ -0,0 +1,533 @@
+and a field of that type. With the unnamed fields mechanism, you can also
+define an unnamed grouping of overlapping fields inline within the `struct`,
+using the `union` keyword:
+
+```rust
+struct S {
+    a: u32,
+    _: union {
+        b: u32,
+        c: f32,
+    },
+    d: u64,
+}
+```
+
+The underscore `_` indicates the absence of a field name; the fields within the
+unnamed union will appear directly with the containing structure. Given a
+struct `s` of this type, code can access `s.a`, `s.d`, and either `s.b` or
+`s.c`. Accesses to `a` and `d` can occur in safe code; accesses to `b` and `c`
+require unsafe code, and `b` and `c` overlap, requiring care to access only the
+field whose contents make sense at the time. As with any `union`, borrows of
+any `union` field borrow the entire union, so code cannot borrow `s.b` and
+`s.c` simultaneously if any of the borrows uses `&mut`.
+
+Conversely, sometimes when defining a `union`, you may want to group multiple
+fields together and make them available simultaneously, with non-overlapping
+storage. You could do so by defining a separate `struct`, and placing an
+instance of that `struct` within the `union`. With the unnamed fields
+mechanism, you can also define an unnamed grouping of non-overlapping fields
+inline within the `union`, using the `struct` keyword:
+
+```rust
+union U {
+    a: u32,
+    _: struct {
+        b: u16,
+        c: f16,
+    },
+    d: f32,
+}
+```
+
+Given a union `u` of this type, code can access `u.a`, or `u.d`, or both `u.b`
+and `u.c`. Since all of these fields can potentially overlap with others,
+accesses to any of them require unsafe code; however, `b` and `c` do not
+overlap with each other. Code can borrow `u.b` and `u.c` simultaneously, but
+cannot borrow any other fields at the same time.
+
+Structs can also contain unnamed structs, and unions can contain unnamed
+unions.
+
+Unnamed fields can contain other unnamed fields. For example:
+
+```rust
+struct S {
+    a: u32,
+    _: union {
+        b: u32,
+        _: struct {
+            c: u16,
+            d: f16,
+        },
+        e: f32,
+    },
+    f: u64,
+}
+```
+
+This structure contains six fields: `a`, `b`, `c`, `d`, `e`, and `f`. Safe code
+can access fields `a` and `f`, at any time, since those fields do not lie
+within a union and do not overlap with any other field. Unsafe code can access
+the remaining fields. This definition effectively acts as the overlap of the
+following three structures:
+
+```rust
+// variant 1
+struct S {
+    a: u32,
+    b: u32,
+    f: u64,
+}
+
+// variant 2
+struct S {
+    a: u32,
+    c: u16,
+    d: f16,
+    f: u64,
+}
+
+// variant 3
+struct S {
+    a: u32,
+    e: f32,
+    f: u64,
+}
+```
+
+## Unnamed fields with named types
+
+An unnamed field may also use a named `struct` or `union` type. For instance:
+
+```rust
+union U {
+    x: i64,
+    y: f64,
+}
+
+struct S {
+    _: U,
+    z: usize,
+}
+```
```

Comment:
Hmm, this reminds me of something
```rust
union Base {
    x: i64,
    y: f64,
}

struct Derived {
    _: Base,
    z: usize,
}
```
except that you need to keep the field order fixed with `#[repr(C)]` to convert pointers, and all the virtual stuff doesn't work. But at least there is some field name reuse!